htlcswitch/resolution_store.go

```diff
@@ -0,0 +1,168 @@
+package htlcswitch
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/contractcourt"
+	"github.com/lightningnetwork/lnd/kvdb"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// resBucketKey is used for the root level bucket that stores the
+	// CircuitKey -> ResolutionMsg mapping.
+	resBucketKey = []byte("resolution-store-bucket-key")
+)
+
+// resolutionStore contains ResolutionMsgs received from the contractcourt. The
+// Switch deletes these from the store when the underlying circuit has been
+// removed via DeleteCircuits. If the circuit hasn't been deleted, the Switch
+// will dispatch the ResolutionMsg to a link if this was a multi-hop HTLC or to
+// itself if the Switch initiated the payment.
+type resolutionStore struct {
+	backend kvdb.Backend
+}
+
+func newResolutionStore(db kvdb.Backend) *resolutionStore {
```

Comment:
Why a new store and not use the same mechanism that is used in the off-chain flow to reliably communicate a resolution from the outgoing link back to incoming link?

Reply:
I think an outgoing link uses the settle/fail filter of the incoming link to do the hand-off in a safe way? I wondered if that same mechanism can be used for on-chain resolutions.

Reply:
After looking into it a little bit, when an on-chain closure is detected and sent to `ChannelArbitrator`, `MarkChannelClosed` is called, which deletes the channel's forwarding packages via `Wipe`. The settle/fail filter is in the outgoing link which would get deleted here. It would be possible to move `Wipe` to the final resolution method `MarkChannelResolved`. I think sending/failing back in the `ChannelArbitrator` would require a resolver so that `MarkChannelClosed` isn't called while we still need the fwdpkg. If the Switch was modified to use pending-closed channels in `reforwardResponses`, then as long as a hand-off occurs the first time, the resolver wouldn't need to do it again as the Switch will reforward responses on start-up. I'm not sure if using pending-closed channels in `reforwardResponses` violates any assumptions though. There is probably a way to make it work, but I really didn't want to add another resolver.

Reply:
I see, sounds not much simpler indeed and definitely riskier. Maybe there is an ideal world where an abstraction layer sits above the on-chain and off-chain logic so that the same pipeline is used in both stages, but that's not the case yet.